### ISSUE TITLE
MissingDataTransferObject

### ISSUE THREAD

**Issue by sidigi** [spatie/data-transfer-object#179]
Partial usage 

idea from this https://github.com/spatie/data-transfer-object/issues/134#issuecomment-675487572

Make it like in laravel Api Resources (MissingValue)

What about values, which was not found in update data transfer object?

```php
class MissingDataTransferObject extends DataTransferObject
{
    public function __construct(...$args)
    {
        if (is_array($args[0] ?? null)) {
            $args = $args[0];
        }

        $r = new ReflectionClass(static::class);

        collect($r->getProperties(ReflectionProperty::IS_PUBLIC))
            ->each(function ($property) use ($args) {
                if (! array_key_exists($property->name, $args)) {
                    $property->setValue($this, $args[$property->name] ?? $this->missingValue($property)?->getMissingInstance() ?? null);

                    return;
                }

                $property->setValue($this, $args[$property->name] ?? null);
            });
    }

    protected function missingValue($property)
    {
        return collect($property->getAttributes())->map(function ($attribute) {
            return $attribute->newInstance();
        })->first(function ($attribute) {
            return $attribute instanceof HasMissingValue;
        });
    }
}
```

```php

namespace Support\Attributes;

#[\Attribute]
class HasMissingValue
{
    public $value;

    public function __construct(string $value)
    {
        $this->value = $value;
    }

    public function getMissingInstance()
    {
        return new $this->value;
    }
}
```

```php
class UpdateCompanyData extends MissingDataTransferObject
{
    #[HasMissingValue(MissingValue::class)]
    public string | MissingValue $title;

    #[HasMissingValue(MissingValue::class)]
    public null | string | MissingValue $description;

    #[HasMissingValue(MissingValue::class)]
    public null | string | MissingValue $logo;
}
```

```php
        dd(new UpdateCompanyData([
            'title' => 'asdfasd',
        ]));
```

```php
Domain\Company\DataTransferObjects\UpdateCompanyData^ {#445
  +title: "asdfasd"
  +description: Support\DataTransferObject\MissingValue^ {#6760}
  +logo: Support\DataTransferObject\MissingValue^ {#6761}
  #exceptKeys: []
  #onlyKeys: []
}
```
```

**Comment by brendt** [comment]
We wouldn't need an attribute, we could evaluate the type hints and check one of the types implement an interface like eg `MissingValue`. That's a little more magic but the less config, the better.

How would this typehint make sense? `null | string | MissingValue` I'd expect `null` values to be casted to the missing value, hence `null` wouldn't be possible.

In essence, this would be a shorthand for supporting the null object pattern. I like it! I'm going to discuss this with some of my colleagues but I'll probably add it soon.

**Comment by sidigi** [comment]
nope

I have a Model. and come of property in my model can be null or strin for example. But, for example i have a value in my table.

Company
- name - required string (in db)
- available_slots- integer or null (in db)

Company
- name: Sony
- available_slots 5

When we nned update only one value in dto - we can not do it, because null is not valid value.

```
new UpdateCompanyData([
            'name' => 'asdfasd',
        ])
```

I wont to reset my available_slots if it doesnt set, and a want have control to set it to null 

```
new UpdateCompanyData([
            'name' => 'asdfasd',
            'available_slots' => null
        ])
```

NULL and ISSET not equal things. Null - is a value, String - is a value, but when we not setting the value - it is a empty value or missing value

**Comment by brendt** [comment]
Regarding the null object pattern I described: I made an implementation last week, and think we won't add it after all, it's too much magic that can already be solved using custom casters.

Regarding your question: one of the major goals of this DTO package is to rely on the type system to prevent bugs and gain insights. Optional values as you describe them go against the strictness this package is trying to promote. Nothing is stopping you though from extending the DTO classes and add your own changes to support it.
